### ISSUE TITLE
bypass PageCache for compact_level0_phase1

### ISSUE THREAD

**Issue by problame** [neondatabase/neon#8184]

**Comment by problame** [comment]
Did some initial scouting work on this:

* the `holes` functionality of compaction (introduced in https://github.com/neondatabase/neon/pull/3597) requires scanning of all keys before scanning all values
  * tl;dr for what holes does: 
    * So the trade-off is that we rather create smaller L1s than creating sparse delta space on top of image layers.
    * And if we're ingesting a lot of data on both sides of the hole, then this is definitely the right trade-off because we will have full-sized L1s on either side.
    * But if we have little L0 data, then we create the small L1s.
    * And all of this is necessary because we have the stupid count_deltas as the trigger for image layer creation.
  * More details in [private Slack DM](https://neondb.slack.com/archives/D049K7HJ9JM/p1721640542312199)
* we _need_ to preserve the holes functionality until we have a better approach for image layer creation at the top
* testing: I can't find any dedicated Rust unit tests. Would be nice to extract the existing logic enough to get coverage for existing behaviors, but that's a lot of work.

**Comment by problame** [comment]
Status update:
 * Initial PR merged #8543 
   * Read the PR description, it contains a detailed rollout strategy.
 * default neon_local, unit tests, regression tests, and Staging are running a validation mode
   * log scraping alerts for staging and prod are configured
   * these will detect validation failures and forward to #on-call-storage-{staging,prod}-stream
 * post-merge of above PR, I ran some manual tests to inspect overhead of validation mode
   * results here: https://github.com/neondatabase/neon/pull/8543#issuecomment-2260710812
 * pre-prod and prod require the following PR to be merged before the release next week:
   * https://github.com/neondatabase/aws/pull/1663
   * If we don't merge it, pre-prod and prod will run in validating mode as well, which we probably don't want yet.

**Comment by problame** [comment]
Status update:

* No validation failures in Staging
* Spot-checked flaky tests dashboard & error messages, no indication that the validation failure is happening there, either.
  * https://github.com/neondatabase/cloud/issues/16404 would be better than spot-checking

Plan / needs decision:

* Validation mode in pre-prod region? Measure CPU impact, see if low enough so we can enable validation mdoe in prod. Probablistic validation?

**Comment by problame** [comment]
Status update: validation mode enabled in pre-prod

[First night's prodlike cloudbench run](https://github.com/neondatabase/cloud/actions/runs/10408266450) had concurrent activity from another benchmark, smearing results: https://neondb.slack.com/archives/C06K38EB05D/p1723797560693199

However, here's the list of dashboards I looked at:

* [Cloudbench metrics](https://neonprod.grafana.net/d/fdbl98ifhoc1se/cloudbench-productionlike-staging?orgId=1&from=1723519506000&to=1723832100000)
* [4GS](https://neonprod.grafana.net/d/d061a222-a79b-4c4b-8e77-16b11545b1fe/pageserver-4-golden-signals?orgId=1&var-datasource=HUNg6jvVk&var-group_by=instance&var-getpage_latency_percentiles=0.99&var-getpage_latency_percentiles=0.999&var-instance=pageserver-8.eu-west-1.aws.neon.build&var-neon_region=eu-west-1&var-basebackup_latency_percentiles=0.99&var-basebackup_latency_percentiles=0.999&from=1723519506746&to=1723832100083)
* [Node Exporter Full](https://neonprod.grafana.net/d/rYdddlPWk/node-exporter-full?orgId=1&from=1723519506000&to=1723832100000&var-DS_PROMETHEUS=HUNg6jvVk&var-job=node-exporter&var-node=pageserver-8.eu-west-1.aws.neon.build&var-diskdevices=%5Ba-z%5D%2B%7Cnvme%5B0-9%5D%2Bn%5B0-9%5D%2B%7Cmmcblk%5B0-9%5D%2B)
* [CPU](https://neonprod.grafana.net/d/fa2dcfb3-fc9a-41f2-8d8f-a91aa6eb3c57/cpu?orgId=1&from=1723519506000&to=1723832100000&var-datasource=HUNg6jvVk&var-instance=pageserver-8.eu-west-1.aws.neon.build&var-region=eu-west-1&var-window=5m)
* [VirtualFile](https://neonprod.grafana.net/d/d0b8e889-39ca-4614-8b00-83db0d0bf868/virtualfile?orgId=1&var-datasource=HUNg6jvVk&var-storage_io_operation_label_operation=All&var-neon_region=All&var-instance=All&var-group_by=instance&var-adhoc=instance%7C%3D%7Cpageserver-8.eu-west-1.aws.neon.build&from=1723519506000&to=1723832100000)
* [/storage disk IOPS](https://neonprod.grafana.net/d/a7350fab-c460-4976-a13c-28f13aec28d6/pageserver-storage-iops-metrics?orgId=1&var-rw=All&var-rw_past=All&var-datasource=HUNg6jvVk&var-ps_nvme1n1=instance%3D~%22pageserver-8.eu-west-1.aws.neon.build%22,neon_service%3D%22pageserver%22,device%3D%22nvme1n1%22&from=1723519506000&to=1723832100000)
* [Memory](https://neonprod.grafana.net/d/c73a3c00-1b4d-474d-ab2a-44df524af8d2/memory?orgId=1&from=1723519506000&to=1723832100000&var-datasource=HUNg6jvVk&var-adhoc=instance%7C%3D%7Cpageserver-8.eu-west-1.aws.neon.build)
* [Fleet overview => memory usage](https://neonprod.grafana.net/d/bee708dc-58ed-4720-b11b-dcbb9f39d3fe/pageserver-fleet-overview?orgId=1&var-datasource=HUNg6jvVk&var-neon_region=All&var-adhoc=instance%7C%3D%7Cpageserver-8.eu-west-1.aws.neon.build&from=1723519506000&to=1723832100000)
* [Time spent doing th ecompactions, scraped from logs](https://neonprod.grafana.net/explore?schemaVersion=1&panes=%7B%22st4%22:%7B%22datasource%22:%22xHHYY0dVz%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22sum%20by%20%28hostname%29%20%28%5Cnsum_over_time%28%7Bhostname%3D%5C%22pageserver-8.eu-west-1.aws.neon.build%5C%22,unit%3D%5C%22pageserver.service%5C%22%7D%7C~%20%60compact_level0_phase1.%2Astats_json%60%20%7C%20regexp%20%60.%2Astats_json%3D%28%3FP%3Cstats_json%3E.%2A%29%60%20%7C%20line_format%20%60%7B%7B.stats_json%7D%7D%60%20%7C%20regexp%20%60new_deltas_size%5C%5C%5C%5C%5C%22:%28%3FP%3Cnew_deltas_size%3E%5C%5Cd%2B%29%60%20%7C%20unwrap%20new_deltas_size%20%5B10m%5D%29%5Cn%29%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22xHHYY0dVz%22%7D,%22editorMode%22:%22code%22,%22hide%22:true%7D,%7B%22refId%22:%22B%22,%22expr%22:%22sum%20by%20%28hostname%29%20%28%5Cnsum_over_time%28%7Bhostname%3D%5C%22pageserver-8.eu-west-1.aws.neon.build%5C%22,unit%3D%5C%22pageserver.service%5C%22%7D%7C~%20%60compact_level0_phase1.%2Astats_json%60%20%7C%20regexp%20%60.%2Astats_json%3D%28%3FP%3Cstats_json%3E.%2A%29%60%20%7C%20line_format%20%60%7B%7B.stats_json%7D%7D%60%20%7C%20regexp%20%60level0_deltas_count%5C%5C%5C%5C%5C%22:%28%3FP%3Clevel0_deltas_count%3E%5C%5Cd%2B%29%60%20%7C%20unwrap%20level0_deltas_count%20%5B10m%5D%29%5Cn%29%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22xHHYY0dVz%22%7D,%22editorMode%22:%22code%22,%22hide%22:true%7D,%7B%22refId%22:%22C%22,%22expr%22:%22sum%20by%20%28hostname%29%20%28%5Cnsum_over_time%28%7Bhostname%3D%5C%22pageserver-8.eu-west-1.aws.neon.build%5C%22,unit%3D%5C%22pageserver.service%5C%22%7D%7C~%20%60compact_level0_phase1.%2Astats_json%60%20%7C%20regexp%20%60.%2Astats_json%3D%28%3FP%3Cstats_json%3E.%2A%29%60%20%7C%20line_format%20%60%7B%7B.stats_json%7D%7D%60%20%7C%20regexp%20%60write_layer_files_micros%5C%5C%5C%5C%5C%22:%28%3FP%3Cwrite_layer_files_micros%3E%5C%5Cd%2B%29%60%20%7C%20unwrap%20write_layer_files_micros%20%5B1h%5D%29%5Cn%29%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22xHHYY0dVz%22%7D,%22editorMode%22:%22code%22,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221723519506000%22,%22to%22:%221723832100000%22%7D%7D%7D&orgId=1)

Preliminary interpretation (compare time range from 0:00 to 8:00, that's where the load happens)

* no noticable CPU / disk IOPS impact
* but compaction iterations take about 2x wall clock time
  * makes sense because validation does about twice the amount of VirtualFile calls, with no concurrency
  * we're not bottlenecking on disk however

![Image](https://github.com/user-attachments/assets/7e37c16b-065f-4366-b5af-3c5efa2bdb7b)

What's the **practical impact**? 2x wall-clock-time-slower compactions means double the wait time on the global semaphore for compactions (assuming that semaphore is the practical throughput bottleneck, which I believe is the case). In other teams, it means we only **achieve half the usual compaction throughput**.

So, is prod compaction throughput bottlenecked on the global semaphore?

We can use the following query to approximate business of the semaphore (%age of tenants waiting for permit):

```
(pageserver_background_loop_semaphore_wait_start_count{instance="pageserver-8.eu-west-1.aws.neon.build",task="compaction"} - pageserver_background_loop_semaphore_wait_finish_count)
/on(instance) pageserver_tenant_states_count{state="Active"}
```

There are some places where we have sampling skew, so, do clamping

```
clamp(
(pageserver_background_loop_semaphore_wait_start_count{task="compaction"} - pageserver_background_loop_semaphore_wait_finish_count)
/on(instance) sum by (instance) (pageserver_tenant_states_count)
, 0, 1)
```


![Image](https://github.com/user-attachments/assets/11287780-3aa2-4048-b486-8242aee80a5b)

the p99.9 instance in that plot looks like this


```
quantile(0.999,
clamp(
(pageserver_background_loop_semaphore_wait_start_count{task="compaction"} - pageserver_background_loop_semaphore_wait_finish_count)
/on(instance) sum by (instance) (pageserver_tenant_states_count)
, 0, 1)
)
```

![Image](https://github.com/user-attachments/assets/b12c465e-cb80-45c2-b416-d42f4c104213)

average like this

```
avg(
clamp(
(pageserver_background_loop_semaphore_wait_start_count{task="compaction"} - pageserver_background_loop_semaphore_wait_finish_count)
/on(instance) sum by (instance) (pageserver_tenant_states_count)
, 0, 1)
)
```


![Image](https://github.com/user-attachments/assets/a625d1f1-ccad-4c6c-8208-a393122cdfab)

**Comment by problame** [comment]
For posterity, there was a Slack thread discussing these results / next steps: https://neondb.slack.com/archives/C033RQ5SPDH/p1723810312846849

**Comment by problame** [comment]
Decision from today's sync meeting:

1. https://github.com/neondatabase/infra/pull/1745
2. Create metric to measure semaphore contention.
3. Table decision for remaining regions until EOW / next week.

**Comment by problame** [comment]
This week, as per discussion thread:

* analyze perf impact in pre-prod (enable new mode, without validation)
* no changes to prod

**Comment by problame** [comment]
Results from pre-prod are looking good.

* [cloudbench results for that time range](https://neonprod.grafana.net/d/fdbl98ifhoc1se/cloudbench-productionlike-staging?orgId=1&from=1724718932000&to=1724858338000&var-datasource=HUNg6jvVk&var-region=eu-west-1&var-percentiles=0.99)
* **[Less wall clock time](https://neonprod.grafana.net/explore?schemaVersion=1&panes=%7B%22st4%22:%7B%22datasource%22:%22xHHYY0dVz%22,%22queries%22:%5B%7B%22refId%22:%22C%22,%22expr%22:%22sum%20by%20%28hostname%29%20%28%5Cnsum_over_time%28%7Bneon_region%3D%5C%22eu-west-1%5C%22,unit%3D%5C%22pageserver.service%5C%22%7D%7C~%20%60compact_level0_phase1.%2Astats_json%60%20%7C%20regexp%20%60.%2Astats_json%3D%28%3FP%3Cstats_json%3E.%2A%29%60%20%7C%20line_format%20%60%7B%7B.stats_json%7D%7D%60%20%7C%20regexp%20%60write_layer_files_micros%5C%5C%5C%5C%5C%22:%28%3FP%3Cwrite_layer_files_micros%3E%5C%5Cd%2B%29%60%20%7C%20unwrap%20write_layer_files_micros%20%5B14h%5D%29%5Cn%29%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22xHHYY0dVz%22%7D,%22editorMode%22:%22code%22,%22hide%22:false,%22step%22:%221h%22%7D%5D,%22range%22:%7B%22from%22:%221724718932162%22,%22to%22:%221724858338297%22%7D%7D%7D&orgId=1)** for the same workload (see screenshot below) => GOOD
* unchanged [memory for that time range](https://neonprod.grafana.net/d/c73a3c00-1b4d-474d-ab2a-44df524af8d2/memory?orgId=1&from=1724718932000&to=1724858338000&var-datasource=HUNg6jvVk&var-adhoc=neon_region%7C%3D%7Ceu-west-1) unchange
* unchanged [CPU for that time range](https://neonprod.grafana.net/d/fa2dcfb3-fc9a-41f2-8d8f-a91aa6eb3c57/cpu?orgId=1&from=1724718932000&to=1724858338000&var-datasource=HUNg6jvVk&var-instance=pageserver-8.eu-west-1.aws.neon.build&var-region=eu-west-1&var-window=5m&viewPanel=3)

![Image](https://github.com/user-attachments/assets/6a6e92d3-b497-47e6-a244-d8dd0e58ea7d)

**Comment by problame** [comment]
Plan:
* Roll the non-validating mode into more prod regions this week.

**Comment by problame** [comment]
Results from rollout shared in [this Slack thread](https://neondb.slack.com/archives/C033RQ5SPDH/p1725530975830579)

tl;dr:

* halved the PS PageCache eviction rate, and stabilized it a lot
* halved the metric "wall clock time spent on compaction / ingested bytes" (see query below)

```
sum by (neon_region) (rate(pageserver_storage_operations_seconds_global_sum{operation="compact",neon_region=~"$neon_region"}[$__rate_interval]))
/
sum by (neon_region) (rate(pageserver_wal_ingest_bytes_received[$__rate_interval] / 1e6))
```
